packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -564,59 +564,53 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
+            const timeout = setTimeout(
+                () => {
+                    clearTimeout(timeout);
+                    abortController?.abort();
+                },
+                snapshotOptions.timeout,
+            );
+        }
+
         const usePost = this.hostPolicy.usePostForTreesLatest;
         // If usePost is false, then make get call for TreesLatest.
         // If usePost is true, make a post call. In case of failure other than the reason for which getWithRetryForTokenRefresh
         // will retry, fallback to get call. In case of error for which getWithRetryForTokenRefresh will retry, let it retry
         // only if it is first failure, otherwise fallback to get call.
-        if (usePost) {
-            try {
-                const odspSnapshot = await this.fetchSnapshotWithTimeout(snapshotOptions, tokenFetchOptions, true);
-                return odspSnapshot;
-            } catch (error) {
-                const errorType = error.errorType;
-                if ((errorType === DriverErrorType.authorizationError || errorType === DriverErrorType.incorrectServerResponse) && tokenFetchOptions.refresh === false) {
-                    throw error;
-                }
-                this.logger.sendErrorEvent({ eventName: "TreeLatest_FallBackToGetRequest" }, error);
-                return this.fetchSnapshotWithTimeout(snapshotOptions, tokenFetchOptions, false);
-            }
-        } else {
-            return this.fetchSnapshotWithTimeout(snapshotOptions, tokenFetchOptions, false);
-        }
-    }
-
-    private async fetchSnapshotWithTimeout(snapshotOptions: ISnapshotOptions, tokenFetchOptions: TokenFetchOptions, usePost: boolean) {
-        assert(snapshotOptions.timeout !== undefined, "Timeout should be provided for setting a limit");
-        const abortController = new AbortController();
-        const timeout = setTimeout(
-            () => {
-                clearTimeout(timeout);
-                abortController.abort();
-            },
-            snapshotOptions.timeout,
-        );
         try {
             const odspSnapshot = await this.fetchSnapshotCore(snapshotOptions, tokenFetchOptions, usePost, abortController);
             return odspSnapshot;
         } catch (error) {
+            const errorType = error.errorType;
+            if ((errorType === DriverErrorType.authorizationError || errorType === DriverErrorType.incorrectServerResponse) && tokenFetchOptions.refresh === false) {
+                throw error;
+            }
             // If the snapshot size is too big and the host specified the size limitation, then don't try to fetch the snapshot again.
-            if (error.errorType === OdspErrorType.snapshotTooBig && this.hostPolicy.snapshotOptions?.mds !== undefined) {
+            if ((error.errorType === OdspErrorType.snapshotTooBig && this.hostPolicy.snapshotOptions?.mds !== undefined) && (this.hostPolicy.summarizerClient !== true)) {
```

Comment:
It's a bit hard to extract intention here.
First, why do we check for mds? If we got snapshotTooBig, we can assert that mds was provided (to check if storage does not return this error code for something unrelated), but really there is no need to check for it, as we have failure in our hands one way or another.

I think this block and next block should be under one big (and same) IF statement - checking for both snapshotTooBig or fetchTimeout. I.e. below we treat them interchangeably, not sure why here we do not do the same.

Reply:
This is because if we get snapshotTooBig error and if the host specified the mds limit, then we don't want to retry without blobs. This is for host perf scenario like Teams as you suggested.

Reply:
I see - we are checking not snapshotOptions, but this.hostPolicy.snapshotOptions.
Maybe put some comments more clearly describing the difference (the original intention of the user).

In general, this is not good practice to use "global" state, it's better to pass both policies explicitly to a function, i.e. hostOptions (what is this.hostPolicy.snapshotOptions& driverOptions (what is snapshotOptions).
Or even better - if we can take logic around line 388 and just move it into this function, such that it only receives hostPolicy, but locally modified the values that it is about to use right here (I have not checked if it's easy / possible). That way code would be more readable / more localized

Reply:
Done as suggested. Now we pass the hostSnapshotOptions from above and separate them from driverSnapshotOptions.